stripe-core/src/main/java/com/stripe/android/core/browser/customtabs/CustomTabsManager.kt

```diff
@@ -0,0 +1,221 @@
+     * @param uri the Uri to be opened.
+     * @param fallback a CustomTabFallback to be used if Custom Tabs is not available.
+     */
+    fun openCustomTab(
+        activity: Activity,
+        uri: Uri,
+        fallback: () -> Unit
+    )
+
+    /**
+     * Warms up the browser for a given URL, so that it loads faster when launched.
+     */
+    fun mayLaunchUrl(url: String): Boolean
+}
+
+/**
+ * Manages the connection to the CustomTabsService.
+ *
+ */
+@RestrictTo(RestrictTo.Scope.LIBRARY_GROUP)
+class CustomTabsManagerImpl @Inject constructor(
+    private val logger: Logger,
+    private val getCustomTabsPackage: GetCustomTabsPackage
+) : CustomTabsManager {
+
+    private var client: CustomTabsClient? = null
+    private var connection: CustomTabsServiceConnection? = null
+    private var session: CustomTabsSession? = null
+    private var serviceBound = false
+
+    /**
+     * Binds the Activity to the CustomTabsService.
+     */
+    override fun onStart(owner: LifecycleOwner) {
+        super.onStart(owner)
+        val context = owner as? Context ?: return
+        if (client != null) return
+        connection = ServiceConnection(object : ServiceConnectionCallback {
+            override fun onServiceConnected(
+                client: CustomTabsClient
+            ) {
+                log("Service connected")
+                this@CustomTabsManagerImpl.client = client
+                this@CustomTabsManagerImpl.client?.warmup(0)
+                session = this@CustomTabsManagerImpl.client?.newSession(CustomTabsCallback())
+            }
+
+            override fun onServiceDisconnected() {
+                log("Service disconnected")
+                client = null
+                session = null
+            }
+        })
+        bindCustomTabService(context)
+    }
+
+    /**
+     * Unbinds the Activity from the CustomTabsService.
+     */
+    override fun onStop(owner: LifecycleOwner) {
+        super.onStop(owner)
+        val context = owner as? Context ?: return
+        if (connection == null) return
+        connection?.let {
+            runCatching {
+                if (serviceBound) {
+                    context.unbindService(it)
+                    serviceBound = false
+                } else {
+                    log("OnStop: service was not bound")
+                }
+            }
+                .onFailure { log("OnStop: couldn't unbind, ${it.stackTraceToString()}") }
+                .onSuccess { log("OnStop: service unbound") }
+        }
+        client = null
+        connection = null
+        session = null
+    }
+
+    /**
+     * Creates or retrieves an exiting CustomTabsSession.
+     *
+     * @return a CustomTabsSession, or null if a connection to the service could not be established.
+     */
+    private fun getSession(): CustomTabsSession? {
+        when {
+            client == null -> session = null
+            session == null -> session = client?.newSession(null)
+        }
```

Comment:
This reads a little confusing, and we should be find with just this:

```suggestion
        session = client?.newSession(null)
```